### PR TITLE
Add feature.T interface

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Signed-off-by: Francesco Guardiani <francescoguard@gmail.com>

Extracted from #97

This PR introduces `feature.T` to restrict the capabilities of `StepFn` to a subset of functionalities supported by the framework. 

A `StepFn` now can fail, err, skip and log but it cannot create a subtest nor do other fancy stuff. This constrain a bit the step capabilities, in order to "force" users to create different feature steps. This also simplifies the assumptions we can do on whether happens in a step fn and allows us to record the stepfn results independently from what go testing is doing. We can also at some point in future reimplement this interface, wrap `*testing.T` or do some other fancy stuff to have proper reporting.

# Changes

- :gift: Add `feature.T` interface
- :broom: Now `StepFn` accepts `feature.T` more than `*testing.T`

**Release Note**

```release-note
Now `StepFn` doesn't accept `*testing.T` anymore, but it requires `feature.T`
```

**Docs**

```docs

```

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
cc @n3wscott

File: pkg/feature/t.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/*
+Copyright 2020 The Knative Authors
```
Comment (n3wscott) [inline pkg/feature/t.go:2]: nit: 2021

File: test/example/config/echo/echo.yaml [hunk 2]
```diff
@@ -15,7 +15,7 @@
 apiVersion: batch/v1
 kind: Job
 metadata:
-  name: {{ .name }}
+  name: echo
```
Comment (n3wscott) [inline test/example/config/echo/echo.yaml:18]: please revert this, I made a fix in the last PR to let the and related stuff not hardcoded

File: test/example/config/echo/echo.go [hunk 3]
```diff
@@ -35,10 +34,9 @@ type Output struct {
 	Message string `json:"msg"`
 }
 
-func Install(name, message string) feature.StepFn {
-	return func(ctx context.Context, t *testing.T) {
+func Install(message string) feature.StepFn {
```
Comment (n3wscott) [inline test/example/config/echo/echo.go:37]: please revert. likely merge cherry-pick overwrite

File: test/example/config/echo/echo_test.go [hunk 4]
```diff
@@ -42,7 +41,7 @@ func Example() {
 	// apiVersion: batch/v1
 	// kind: Job
 	// metadata:
-	//   name: echo-123
```
Comment (n3wscott) [inline test/example/config/echo/echo_test.go:45]: revert

File: test/example/config/echo/echo_test.go [hunk 5]
```diff
@@ -27,7 +27,6 @@ func Example() {
 		"ko://knative.dev/reconciler-test/test/example/cmd/echo": "uri://a-real-container",
 	}
 	cfg := map[string]interface{}{
-		"name":      "echo-123",
```
Comment (n3wscott) [inline test/example/config/echo/echo_test.go:30]: revert

File: test/example/echo_feature.go [hunk 6]
```diff
@@ -20,41 +20,41 @@ import (
...
 	kubeclient "knative.dev/pkg/client/injection/kube/client"
+
 	"knative.dev/reconciler-test/pkg/environment"
 	"knative.dev/reconciler-test/pkg/feature"
 	"knative.dev/reconciler-test/pkg/k8s"
 	"knative.dev/reconciler-test/test/example/config/echo"
 )
 
 func EchoFeature() *feature.Feature {
-	f := new(feature.Feature)
```
Comment (n3wscott) [inline test/example/echo_feature.go:35]: revert

File: test/example/echo_feature.go [hunk 7]
```diff
@@ -20,41 +20,41 @@ import (
...
 	})
 
 	f.Alpha("pull logs off a pod").
 		Must("the echo pod must echo our message",
-			func(ctx context.Context, t *testing.T) {
+			func(ctx context.Context, t feature.T) {
 				env := environment.FromContext(ctx)
 				client := kubeclient.Get(ctx)
 
-				log, err := k8s.WaitForJobTerminationMessage(client, env.Namespace(), name, time.Second, 30*time.Second)
```
Comment (n3wscott) [inline test/example/echo_feature.go:57]: revert

File: test/example/echo_feature.go [hunk 8]
```diff
@@ -76,25 +76,12 @@ func EchoFeature() *feature.Feature {
...
-		Should("An example of a SHOULD", func(ctx context.Context, t *testing.T) {
+		Should("An example of a SHOULD", func(ctx context.Context, t feature.T) {
 			t.Log("ran inside of a SHOULD")
 		})
 
 	return f
 }
-
-// EchoFeatureSet makes a feature set out of a few EchoFeatures for testing.
-func EchoFeatureSet() *feature.FeatureSet {
```
Comment (n3wscott) [inline test/example/echo_feature.go:90]: revert

**Comment by slinkydeveloper** [comment]
Fixed cherry-pick issues, sorry for this

**Review by n3wscott — APPROVED** [review]
/lgtm
/approve

**Pull request merged**